### DISCUSSION TITLE
How to make private-lib work with dynamic library injection?

### DISCUSSION THREAD

**Discussion by osevan** [netblue30/firejail#6281] [Q&A]
Hi community, 

im using firejal since 2016 amd love every improvement.

Today i want ask a question about one of my favourite functions private-lib:

Im maintaining seperate glibc under subfolder of default debian sid library path and want use these librarys - because my apps compiled already dynamicly against them - inside fj containers private-lib function.

How i can tell Firejail for my seperate glibc directory instead of system defaults directory?

If this not existing yet, could we make argument as private-lib-path=/usr/lib/mypath? 

Thanks and 
Best regards

**Comment by ghost** [comment]
Hi @osevan.

You might be already aware of the below remarks regarding `private-lib`, but I'll just put them in anyway in case other users read this.

- check if your OS builds FJ with private-lib support. It will show in the `firejail --version` output; if not you'll have to build FJ from source and use the `--enable-private-lib` flag in configure

- /etc/firejail/firejail.config
```sh
$ grep private-lib /etc/firejail/firejail.config
# Enable or disable private-lib feature, default disabled
private-lib yes
```

> glibc under subfolder of default debian sid library path

I think in that case (subfolder) you're good to go. Doesn't hurt to run a few quick tests to make sure, before you start adding private-lib stuff in your .local overrides. And beware of some potential pitfalls, see the below examples:

```sh
$ firejail --ignore=quiet --private-lib -- echo "foo"
...
Standard C library installed in 5.62 ms
Firejail libraries installed in 5.44 ms
Program libraries installed in 42.01 ms
Installed 26 libraries and 2 directories
...
/bin/bash: error while loading shared libraries: libreadline.so.8: cannot open shared object file: No such file or directory
```

```sh
$ firejail --ignore=quiet --private-lib=libreadline.so.* -- echo "foo"
...
Standard C library installed in 5.62 ms
Firejail libraries installed in 5.44 ms
Program libraries installed in 42.01 ms
Installed 29 libraries and 2 directories
...
foo
```

Feel free to ask away, I use private-lib too in several profiles.

HTH

**Reply by osevan** [reply]
I know already --enable-ptivate-lib at configute level.


I want to tell firejail to use different libc.so instead of system libc.

Plus my own directory with different libc with additional other libs...

**Reply by ghost** [reply]
The snag here _might_be that Firejail uses lists in [fs_lib.c](https://github.com/netblue30/firejail/blob/898273ac8ec3e4366e81dbdf4826b644aff28f09/src/firejail/fs_lib.c) and [fs_lib2.c](https://github.com/netblue30/firejail/blob/898273ac8ec3e4366e81dbdf4826b644aff28f09/src/firejail/fs_lib2.c). I'm not sure if there's some precedence that can ensure your preferred libc.so will be used - for example by placing it in /usr/local/... somewhere.

So what you want _might_ need patching the code (intensively). And it wouldn't surprise me if there was little to no interest in doing so. The private-lib option was poised to be dropped all together in the past (see https://github.com/netblue30/firejail/issues/5190).

That being said, have you tried LD_PRELOAD yet (Inspired by https://github.com/netblue30/firejail/discussions/5512)?

In profile(s):

```
env LD_PRELOAD=/path/to/my/different/libc.so
private-lib /path/to/my/different/libc.so
```

TL;DR you'll need to experiment...

**Reply by rusty-snake** [reply]
> That being said, have you tried LD_PRELOAD yet

`LD_LIBRARY_PATH`

> TL;DR you'll need to experiment...

TL;DR: Not a firejail issue. Just configure your ld.so.

**Reply by osevan** [reply]
How can i configure ld.so only for firejail apps?

**Reply by osevan** [reply]
Btw, i have niy only different libc.so i have created seperated subfolders under libs and copy every library in.


But i will test something

**Comment by osevan** [comment]
I think setting temporart ld_library_path should do this job.


But, will private-lib tool follow this path before binary start and symlinking or copying inside virtualfs?

**Comment by osevan** [comment]
Still need answers.

Because firejail doesnt follow different glibc folder in private-lib flag inside profile. 


I want inject every library path with new glibc path.
